pkg/webhooks/config.go

```diff
@@ -0,0 +1,336 @@
+package webhooks
+
+import (
+	capm3apis "github.com/metal3-io/cluster-api-provider-metal3/api/v1beta1"
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	DefaultMachineMutatingHookPath                     = "/mutate-machine-openshift-io-v1beta1-machine"
+	DefaultMachineValidatingHookPath                   = "/validate-machine-openshift-io-v1beta1-machine"
+	DefaultMachineSetMutatingHookPath                  = "/mutate-machine-openshift-io-v1beta1-machineset"
+	DefaultMachineSetValidatingHookPath                = "/validate-machine-openshift-io-v1beta1-machineset"
+	DefaultMetal3RemediationMutatingHookPath           = "/mutate-infrastructure-cluster-x-k8s-io-v1beta1-metal3remediation"
+	DefaultMetal3RemediationValidatingHookPath         = "/validate-infrastructure-cluster-x-k8s-io-v1beta1-metal3remediation"
+	DefaultMetal3RemediationTemplateMutatingHookPath   = "/mutate-infrastructure-cluster-x-k8s-io-v1beta1-metal3remediationtemplate"
+	DefaultMetal3RemediationTemplateValidatingHookPath = "/validate-infrastructure-cluster-x-k8s-io-v1beta1-metal3remediationtemplate"
+
+	defaultWebhookConfigurationName = "machine-api"
+	defaultWebhookServiceNamespace  = "openshift-machine-api"
+	defaultWebhookServicePort       = 443
+	// Name and port of the webhook service pointing to the machineset-controller container
+	defaultWebhookServiceName = "machine-api-operator-webhook"
+	// Name and port of the webhook service pointing to the machine-controller / provider container
+	// Used by metal3 remediation webhooks in the CAPBM image
+	defaultProviderWebhookServiceName = "machine-api-operator-machine-webhook"
+)
+
+var (
+	// webhookFailurePolicy is ignore so we don't want to block machine lifecycle on the webhook operational aspects.
+	// This would be particularly problematic for chicken egg issues when bootstrapping a cluster.
+	webhookFailurePolicy = admissionregistrationv1.Ignore
+	webhookSideEffects   = admissionregistrationv1.SideEffectClassNone
+)
+
+// NewValidatingWebhookConfiguration creates a validation webhook configuration with configured Machine and MachineSet webhooks
+func NewValidatingWebhookConfiguration() *admissionregistrationv1.ValidatingWebhookConfiguration {
+	validatingWebhookConfiguration := &admissionregistrationv1.ValidatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: defaultWebhookConfigurationName,
+			Annotations: map[string]string{
+				"service.beta.openshift.io/inject-cabundle": "true",
+			},
+		},
+		Webhooks: []admissionregistrationv1.ValidatingWebhook{
+			MachineValidatingWebhook(),
+			MachineSetValidatingWebhook(),
+			Metal3RemediationValidatingWebhook(),
+			Metal3RemediationTemplateValidatingWebhook(),
```

Comment:
Do we want to deploy this on all platforms? Every machine action will check for a webhook that is unlikely to be there in more cases than it is likely to be there.

Perhaps these should be applied as a separate `ValidatingWebhookConfiguration` rather than joined with the other Machine API ones?

Reply:
Maybe I misunderstand your concern, but I don't think machine actions will check these new webhooks. Part of the webhook configuation is the kind for which they are registered 🤔 E.g.
```
				Rule: admissionregistrationv1.Rule{
					APIGroups:   []string{capm3apis.GroupVersion.Group},
					APIVersions: []string{capm3apis.GroupVersion.Version},
					Resources:   []string{"metal3remediations"},
				},
```

Reply:
Right ok, I thought we were adding new machine webhooks which would have been more impactful.

I am still concerned we are setting a precedent though, would it not be better for each subsystem to maintain their own MWC/VWC rather than just adding more and more config to the existing Machine one?

Reply:
What do you mean with "adding more and more config to the existing Machine one"?

The new MWC / VWC are not related to Machines at all. They are independent MWC / VMC resources. That's actually why I moved  all configs into `config.go`, instead of adding the new ones to `machine_webhook.go` (which was done for machineset configs before), to avoid confusion. See 1st commit.

The only thing which makes this related to machines is, that the container name of the provider is `machine-controller`, so some names of new ports/mounts also use `machine`, to make it clear to which container they belong.

Reply:
I mean that this Pr is adding additional `Webhooks` to an existing `ValidatingWebhookConfiguration`, so when I look at `oc get validatingwebhookconfigurations` I will see `machine-api` with 4 webhooks defined within it.

I think for separation of concerns it's likely better if the existing `machine-api` webhook configuration retains it's original 2 webhooks and we create a separate `machine-api-metal3-remediation` ValidatingWebhookConfiguration that handles the `Metal3Remediation...` webhooks

Reply:
Oh sorry, now I got it. Done!